data_utils/ds_addmult_mod10_helpers.py

```diff
@@ -155,20 +207,36 @@ def eval_callback_mod10(model, in_vocab, split, datasets, eval_batch_size=32, du
             label = batch['labels'].long().to(device)
 
             # Get the model's predictions
-            outputs = model(input, input_len, None, None, None)
-            _, predicted_class = torch.max(outputs, dim=1)
-
-            # Update accuracy counters
-            total_count += label.size(0)
-            correct_count += (predicted_class == label).sum().item()
-
-            if (dump):
-                strings = batch['string']
-                label = label.cpu().numpy()
-                predicted_class = predicted_class.cpu().numpy()
-                for idx in range(len(label)):
-                    if (label[idx] != predicted_class[idx]):
-                        dump_file.write(strings[idx] + '\n')
+            if has_token_labels: # TransformerLM
+                prefix_sos = (torch.ones((input.shape[0],1)).long() * model.encoder_sos).to(device)
+                input = torch.cat((prefix_sos, input), dim=1) # add SOS token
+                input_len += 1
+
+                outputs = model(input, input_len)
+                _, preds = torch.max(outputs['data'], dim=2)
+
+                labels_cmp = torch.cat((label.T, prefix_sos), dim=1)
+                labels_match = (preds == labels_cmp) * (input != 0) * (labels_cmp != 0)
+
+                # Update accuracy counters
+                total_count += batch['labels_len'].sum().item()
+                correct_count += labels_match.sum().item()
```

Comment:
Was there a reason this is not implemented inside of `eval_callback_mod10_lm`? Just curious / wanted to get your thoughts in case we future refactor (and use this to aid refactor design)!

Reply:
Yes, because our intermediate supervision head is a hybrid head which uses a classification objective for the per-token labels, not language modeling

If you're using language modeling, that means you're actually inserting the intermediate solutions within the input text - doesn't quite make sense

Reply:
Ah makes sense, thank you for clarification.